react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts

```diff
@@ -0,0 +1,65 @@
+import { UnparsedItem, ItemTree } from './useTableTreeCheckboxes'
+
+/**
+ * Return new state with items toggled
+ * @param state
+ * @param item
+ */
+export function getToggledState(
+  state: Array<ItemTree>,
+  item: ItemTree
+): Array<ItemTree> {
+  const stateIncludesItem = state.some(row => row.id === item.id)
+
+  const bulkFilter = (row: ItemTree) => !getFlat(item).some(equalsId(row))
+
+  const filter = (row: ItemTree) => row.id !== item.id
+
+  const bulkCheck = (state: Array<ItemTree>, item: ItemTree) => {
+    const reduction = (acc: Array<ItemTree>, item: ItemTree) =>
+      acc.some(equalsId(item)) ? acc : [...acc, item]
+
+    return [...state, ...getFlat(item)].reduce(reduction, [])
+  }
+
+  if (stateIncludesItem) {
+    return item.children ? state.filter(bulkFilter) : state.filter(filter)
+  } else {
+    return item.children ? bulkCheck(state, item) : [...state, item]
+  }
+}
+
+/**
+ * Represents a tree section on a single array.
+ */
+export function getFlat(tree: ItemTree, arr: Array<ItemTree> = []) {
+  arr.push(tree)
+  tree.children && tree.children.forEach(child => getFlat(child, arr))
```

Comment:
Guards are great when returning a value. For this, it might be too much, don't you think?